src/tribler/core/components/restapi/rest/utils.py

```diff
@@ -66,3 +73,42 @@ def fix_unicode_array(arr):
             new_arr.append(item)
 
     return new_arr
+
+
+def shorten(s: Optional[str], width: int = 50, placeholder='[...]', cut_at_the_end: bool = True):
+    """ Shorten a string to a given width.
+        Example: shorten('hello world', 5) -> 'hello[...]'
+    """
+    if s and len(s) > width:
+        return f'{s[:width]}{placeholder}' if cut_at_the_end else f'{placeholder}{s[len(s) - width:]}'
+    return s
+
+
+def format_frames(frame: Optional[FrameType], file_width: int = 50, value_width: int = 100) -> Iterator[str]:
+    """ Format a stack trace."""
+    while frame:
+        filename = shorten(frame.f_code.co_filename, width=file_width, cut_at_the_end=False)
+        header = f"{filename}:{frame.f_lineno} {frame.f_code.co_name}"
+        local = ''
+        for key, value in list(frame.f_locals.items()):
+            value = shorten(repr(value), width=value_width)
```

Comment:
Taking the `repr` of an unknown value can trigger an exception; I think we should add a try/except block here. Something like:
```python
try:
    value = repr(value)
except Exception as e:
    value = f'<exception when taking repr: {e.__class__.__name__}: {e}>'
value = shorten(value, width=value_width)    
```

Reply:
Could you provide an example of a dangerous call?

Reply:
It is expected for the `__repr__` method of any object to return a string and don't raise any exception. With this, a `__repr__` method is just a usual method and can raise an exception due to a bug in its code.
```python
class AClassWithABadRepr1:
    def __repr__(self):
        1/0
```

In Tribler, we had a case of incorrect implementation of the `alert.__str__()` method that raises `UnicodeDecodeError` (#5740). In this case, the `alert.__repr__()` implementation does not raise an error, but it is possible to discover a similar problem later with an incorrect `__repr__` implementation of some class.

As another example, historically, some `repr` methods of container classes could not handle situations when the container itself was added as its item (or as an item of another item, etc.) and raised `RecursionError`. Now Python standard structures like lists and dicts handle this case correctly, but it is possible that some user-defined classes still have this problem.

[Some blog posts](https://blog.ionelmc.ro/2020/01/20/is-there-anything-safe-in-python/) describe the problems with the repr of arbitrary objects.

Python issue tracker has an issue for a case like ours: [traceback.FrameSummary does not handle exceptions from `repr()`](https://bugs.python.org/issue39228). They decided not to catch these exceptions because Python follows the ["fail fast"](https://en.wikipedia.org/wiki/Fail-fast) approach:
> Bugs should be reported, not masked. I don't think that FrameSummary should be an exception to this.

But they agree that the case is real, and the exception is possible. We don't follow the "fail fast" approach in Tribler, so we don't want Tribler to crash because a user opened a debug window.

Also, `unittest.util` has a `safe_repr` function, and probably for a reason.

Reply:
I'm not entirely sure about your statement that 'we don't follow the "fail fast" approach in Tribler'. In the previous dev meeting, the majority of the developers voted to maintain the current approach, which is a form of "fail fast".

However, I do agree with your argument and appreciate the clear examples you've provided.

I have now implemented your suggestion.